[TubeTuGraz] Add new extractor

Ferdi265 opened the pull request:
### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
  - This is a port of [Youtube-DL#29160](https://github.com/ytdl-org/youtube-dl/pull/29160) to yt-dlp
  - [Youtube-DL#26778](https://github.com/ytdl-org/youtube-dl/pull/26778) is similar, but with less features. I have talked to @ruru4143, and he has suggested creating a new PR for this and closing the old one
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
  - I have written the bulk of the extractor code from scratch
- [X] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)
  - Some of the code is inspired by @ruru4143's PR, who released it under the Unlicense according to that post. 

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [X] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Add support for extracting videos from [tube.tugraz.at](https://tube.tugraz.at), the lecture video site from Graz University of Technology.

Features:

- extract videos
- extract playlists
- extract restricted videos and playlists if login details are provided
- attempt fallback formats if metadata cannot be downloaded
- formats are assigned human-readable `format_id`s
- "presentation" formats are preferred over "presenter" formats

Reading the original PR ([Youtube-DL#29160](https://github.com/ytdl-org/youtube-dl/pull/29160)) might be useful, since the comments to that PR contain some discussion about the detailed design.

The original PR got stalled because of coding style issues, although the maintainer of Youtube-DL refused to elaborate on what exactly was the issue. I tried to follow the coding style guidelines as closely as possible, and I can't really see an issue.

I skimmed the YT-DLP coding style guidelines as well, and it appears nothing has changed compared to Youtube-DL's. If I misread that, and the coding style did change, please mention it so I can fix it.

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+import re
+
+
+def try_get_or(src, getters, expected_type=None):
+    for getter in getters:
+        v = try_get(src, getter, expected_type)
+        if v is not None:
+            return v
```
try_get already allows multiple getters

fyi: You can use traverse_obj to make it even simpler

Ferdi265 replied:
Thanks, I'll look into it

Ferdi265 replied:
fixed, thank you

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+    IE_DESC = 'tube.tugraz.at'
+    IE_NAME = 'TubeTuGraz'
+
+    _VALID_URL = r"""(?x)^
```
Use singlequotes everywhere

Ferdi265 replied:
will do

Ferdi265 replied:
done

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+    }
+
+    _LOGIN_REQUIRED = False
+    _NETRC_MACHINE = None
```
netrc_machine must be set when login is implemented

Ferdi265 replied:
thanks, chose a machine identifier in the same style as of all of the others

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+        }
+    }
+
+    _LOGIN_REQUIRED = False
```
unused

Ferdi265 replied:
ahh, ok, will remove. I initially thought that is somehow part of the extractor API, but it seems it is not.

Ferdi265 replied:
fixed

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+            (?P<series>browse\.html\?series=)|
+            (?P<episode>watch.html\?id=)
+        )(?P<id>[0-9a-fA-F]{8}-(?:[0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12})
+    $"""
```
$ unnecessary

Ferdi265 replied:
fixed

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+        if username is None:
+            return
+
+        result = self._download_webpage_handle(
```
_request_webpage

Ferdi265 replied:
done for this request, don't know exactly how to do that for the first request

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+            _, login_page_handle = result
+
+        result = self._download_webpage_handle(
+            login_page_handle.url, None,
```
Use `.geturl()`

Ferdi265 replied:
done

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+        if username is None:
+            return
+
+        result = self._download_webpage_handle(
+            self._LOGIN_URL, None,
+            note='downloading login page',
+            errnote='unable to fetch login page',
+            fatal=False)
+        if result is False:
+            return
+        else:
+            _, login_page_handle = result
```
Are you sure this request is even necessary? POSTing to `_LOGIN_URL` should redirect automatically

Ferdi265 replied:
POSTing to `_LOGIN_URL` does not work. It seems the page relies on specifically the referer from the redirected login page to be present

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+            errnote='unable to log in',
+            fatal=False,
+            data=urlencode_postdata({
+                b"lang": "de",
```
`b''` is unnecessary

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+            headers={
+                "referer": login_page_handle.url
+            })
+        if result is False:
+            return
```
this should throw error, no?

Ferdi265 replied:
I'd rather have the extractor try to continue after some kind of failed login, since many videos and even some videos that require login for metadata can still be downloaded without being logged in.

But I must say I don't completely understand the mechanics of the request functions in the YT-DLP utils library, so I don't know exactly how to handle this.

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+        self._login()
+
+    def _real_extract(self, url):
+        match = re.match(self._VALID_URL, url)
```
self._match_valid_url

Ferdi265 replied:
done

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+    def _real_extract(self, url):
+        match = re.match(self._VALID_URL, url)
+
+        id = match.group('id')
+        if match.group('series') is not None:
+            return self._extract_series(id)
+        elif match.group('episode') is not None:
+            return self._extract_episode(id)
+        else:
+            raise ExtractorError("no video found on page")
+
```
better to split this into 2 extractors

Ferdi265 replied:
I don't know exactly. Both extractor paths share a lot of code and splitting the two would probably make the code only marginally simpler.

If this should be split, Series should inherit from the Episode extractor, since it uses a lot of the code from it.
Is there precedence in YT-DLP for episodes and playlists being split into multiple extractors?

pukkandan replied:
create a base class with the common functions and inherit both extractors from it. The base class is conventionally called `SomeSiteBaseIE`

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+
+    def _extract_series(self, id):
+        series_info_data = self._download_json(
+            self._API_SERIES, None,
```
```suggestion
            self._API_SERIES, id,
```

Ferdi265 replied:
fixed, should now pass the video id in on each metadata download call

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+            series_info_data,
+            lambda x: x["catalogs"][0]["http://purl.org/dc/terms/"]) or {}
+
+        if len(series_info) == 0:
+            self.report_warning(
+                "failed to download series metadata: "
+                + "authentication required or series does not exist", id)
```
why warning? shouldn't this be an error?

Also, 
1. don't use `+`
2. `if not series_info`

Ferdi265 replied:
if the warning message is too long, is it okay if it overshoots the 80 character limit in order to not use a `+`?

pukkandan replied:
Yes

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+            "_type": "playlist",
+            "id": id,
+            "title": title,
+            "entries": [self._extract_episode_from_info(episode_info)
+                        for episode_info in episodes_info]
```
use a generator or even better, pagedlist (with pagesize=1)

pukkandan replied:
On second thought, it is best if this returns `url_result`s

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+    def _extract_formats(self, format_info, format_types):
+        PREFERRED_TYPE = "presentation"
+
+        url = try_get(format_info, lambda x: x["url"])
+        type = try_get(format_info, lambda x: x["type"]) or "unknown"
+        transport = try_get(format_info, lambda x: x["transport"]) or "https"
```
for single key, use `.get`

Ferdi265 replied:
fixed

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+        url = try_get(format_info, lambda x: x["url"])
+        type = try_get(format_info, lambda x: x["type"]) or "unknown"
+        transport = try_get(format_info, lambda x: x["transport"]) or "https"
+        audio_bitrate = try_get(format_info, lambda x: x["audio"]["bitrate"])
+        video_bitrate = try_get(format_info, lambda x: x["video"]["bitrate"])
```
```suggestion
        audio_bitrate = try_get(format_info, lambda x: x["audio"]["bitrate"] / 1000)
        video_bitrate = try_get(format_info, lambda x: x["video"]["bitrate"] / 1000)
```

Ferdi265 replied:
done, with your suggested change

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+        if transport == "https":
+            formats = [{
+                "url": url,
+                "tbr": bitrate,
```
tbr will be automatically set

Ferdi265 replied:
removed

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+        type = str(type).replace("/delivery", "")
+        transport = str(transport).lower()
+
+        if isinstance(audio_bitrate, int):
+            audio_bitrate = audio_bitrate / 1000
+        if isinstance(video_bitrate, int):
+            video_bitrate = video_bitrate / 1000
+        if isinstance(audio_bitrate, int) and isinstance(video_bitrate, int):
+            bitrate = audio_bitrate + video_bitrate
+        else:
+            bitrate = None
```
see above comments

Ferdi265 replied:
removed, as per above change

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+                "tbr": bitrate,
+                "abr": audio_bitrate,
+                "vbr": video_bitrate,
+                "framerate": framerate,
```
fps

Ferdi265 replied:
done

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+                "abr": audio_bitrate,
+                "vbr": video_bitrate,
+                "framerate": framerate,
+                "resolution": resolution,
```
instead, set height/width if possible

Ferdi265 replied:
thanks, set to width and height from `parse_resolution`.
in case `parse_resolution` fails, should width and height be None or unset?

currently they will be None, but I don't know if that will cause problems

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+            formats = []
+
+        for format in formats:
+            format["preference"] = preference
```
`preference` should generally not be set. Can you explain the diff b/w the two types of formats? From my understanding, `source_preference` makes more sense?

Ferdi265 replied:
tube.tugraz.at has many lectures that have both a camera stream from the lecture hall ("PRESENTER"), showing the lecturer and the blackboard, as well as a video stream from the connected computer ("PRESENTATION"), usually showing a powerpoint presentation or similar.

The majority of lecture videos uses only PRESENTATION streams, or has a PRESENTER stream that doesn't really show anything useful, and most lecturers that use the blackboard don't use a PRESENTATION stream at all.

These are two separate video streams that can be watched side-by-side on the website if both are present. As far as I understand it, it's hard to map this to YT-DLP's model, so making a choice of which stream to choose by default is necessary if consistent behaviour is desired.

Choosing the best PRESENTATION stream by default is most likely the most sane option, because it will almost always deliver a video that actually contains the content you want to watch. In cases where you want a specific stream, you can still choose.

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+
+                    formats.extend(dash_formats)
+
+    def _gen_format_id(self, format, type, transport, format_types):
+        format_types[type][transport] += 1
+        index = format_types[type][transport]
+
+        vbr = format.get("vbr")
+        abr = format.get("abr")
+        tbr = format.get("tbr") or ((vbr or 0) + (abr or 0))
+        vcodec = format.get("vcodec")
+        acodec = format.get("acodec")
+        if acodec == "none" or (vbr and not abr):
+            pre_index = "vbr"
+            index = int(tbr)
+        elif vcodec == "none" or (abr and not vbr):
+            pre_index = "abr"
+            index = int(tbr)
+        elif tbr:
+            pre_index = "br"
+            index = int(tbr)
+        else:
+            pre_index = ""
+
+        format_id = "%s-%s-%s%d" % (type, transport, pre_index, index)
+        format["format_id"] = format_id
```
unnecessary

Ferdi265 replied:
IMO this is necessary if users actually want to know what they are downloading. (see the note about `preference`)

Here is a comparison of formats with and without `_gen_format_id()` (for video `https://tube.tugraz.at/paella/ui/watch.html?id=6b78b5ef-605d-423c-9b6b-5e6260121f26`):

With format ids:
```
presenter-dash-abr129      m4a        audio only  [eng] DASH audio,  129k, m4a_dash container, mp4a.40.2@129k (44100Hz), ~83.55GiB
presenter-dash-abr162      m4a        audio only  [eng] DASH audio,  162k, m4a_dash container, mp4a.40.2@162k (44100Hz), ~104.92GiB
presenter-dash-abr196      m4a        audio only  [eng] DASH audio,  196k, m4a_dash container, mp4a.40.2@196k (44100Hz), ~126.53GiB
presenter-dash-vbr2134     mp4        768x432     DASH video, 2134k, mp4_dash container, avc1.4d401e@2134k, 30fps, video only, ~1.35TiB
presenter-dash-vbr2150     mp4        1280x720    DASH video, 2150k, mp4_dash container, avc1.4d401f@2150k, 30fps, video only, ~1.36TiB
presenter-dash-vbr2267     mp4        1920x1080   DASH video, 2267k, mp4_dash container, avc1.4d4028@2267k, 30fps, video only, ~1.43TiB
presenter-https-br2134     mp4        768x432     2134kvideo@2004k, 30fps, audio@129k, ~1.35TiB
presenter-hls-br2134       mp4        768x432     2134kvideo@2134k, audio@  0k, ~1.35TiB
presenter-https-br2150     mp4        1280x720    2150kvideo@1988k, 30fps, audio@162k, ~1.36TiB
presenter-hls-br2150       mp4        1280x720    2150kvideo@2150k, audio@  0k, ~1.36TiB
presenter-https-br2267     mp4        1920x1080   2267kvideo@2071k, 30fps, audio@196k, ~1.43TiB
presenter-hls-br2267       mp4        1920x1080   2267kvideo@2267k, audio@  0k, ~1.43TiB
presentation-dash-vbr270   mp4        512x320     DASH video,  270k, mp4_dash container, avc1.4d4015@ 270k, 30fps, video only, ~174.28GiB
presentation-dash-vbr481   mp4        852x532     DASH video,  481k, mp4_dash container, avc1.4d401f@ 481k, 30fps, video only, ~310.53GiB
presentation-dash-vbr699   mp4        1280x800    DASH video,  699k, mp4_dash container, avc1.4d4020@ 699k, 30fps, video only, ~451.63GiB
presentation-https-vbr270  mp4        512x320      270k, 30fps
presentation-https-vbr481  mp4        852x532      481k, 30fps
presentation-https-vbr699  mp4        1280x800     699k, 30fps
presentation-hls-br270     mp4        512x320      270kvideo@ 270k, audio@  0k, ~174.28GiB
presentation-hls-br481     mp4        852x532      481kvideo@ 481k, audio@  0k, ~310.53GiB
presentation-hls-br699     mp4        1280x800     699kvideo@ 699k, audio@  0k, ~451.63GiB
```

Without format ids:
```
p0aa0br129446   m4a        audio only  [eng] DASH audio,  129k, m4a_dash container, mp4a.40.2@129k (44100Hz), ~83.55GiB
p0aa0br162563   m4a        audio only  [eng] DASH audio,  162k, m4a_dash container, mp4a.40.2@162k (44100Hz), ~104.92GiB
p0aa0br196045   m4a        audio only  [eng] DASH audio,  196k, m4a_dash container, mp4a.40.2@196k (44100Hz), ~126.53GiB
p0va0br2134088  mp4        768x432     DASH video, 2134k, mp4_dash container, avc1.4d401e@2134k, 30fps, video only, ~1.35TiB
p0va0br2150658  mp4        1280x720    DASH video, 2150k, mp4_dash container, avc1.4d401f@2150k, 30fps, video only, ~1.36TiB
p0va0br2267088  mp4        1920x1080   DASH video, 2267k, mp4_dash container, avc1.4d4028@2267k, 30fps, video only, ~1.43TiB
6               mp4        768x432     2134kvideo@2004k, 30fps, audio@129k, ~1.35TiB
2134            mp4        768x432     2134kvideo@2134k, audio@  0k, ~1.35TiB
8               mp4        1280x720    2150kvideo@1988k, 30fps, audio@162k, ~1.36TiB
2150            mp4        1280x720    2150kvideo@2150k, audio@  0k, ~1.36TiB
10              mp4        1920x1080   2267kvideo@2071k, 30fps, audio@196k, ~1.43TiB
2267            mp4        1920x1080   2267kvideo@2267k, audio@  0k, ~1.43TiB
p0va0br270028   mp4        512x320     DASH video,  270k, mp4_dash container, avc1.4d4015@ 270k, 30fps, video only, ~174.28GiB
p0va0br481131   mp4        852x532     DASH video,  481k, mp4_dash container, avc1.4d401f@ 481k, 30fps, video only, ~310.53GiB
p0va0br699756   mp4        1280x800    DASH video,  699k, mp4_dash container, avc1.4d4020@ 699k, 30fps, video only, ~451.63GiB
15              mp4        512x320      270k, 30fps
16              mp4        852x532      481k, 30fps
17              mp4        1280x800     699k, 30fps
270             mp4        512x320      270kvideo@ 270k, audio@  0k, ~174.28GiB
481             mp4        852x532      481kvideo@ 481k, audio@  0k, ~310.53GiB
699             mp4        1280x800     699kvideo@ 699k, audio@  0k, ~451.63GiB
```

It's almost impossible to discern whether a format is a PRESENTATION or PRESENTER format without labelling them.

Additionally, some transports seem to be slower to download than others (the full video downloads over https usually go far quicker than the chunked HLS or DASH downloads, at least in my experience).

Also, please disregard the sizes listed next to the formats. I know they are completely wrong, I'm currently trying to figure out what causes them to be off.

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+
+                    formats.extend(hls_formats)
+            if "dash" not in format_types[type]:
+                dash_page = self._download_webpage_handle(
+                    mpd_url, None,
+                    note="guessing location of %s DASH manifest" % type,
+                    errnote=False,
+                    fatal=False)
```
unnecessary. If URL doesnt connect, _extract_mpd_formats will return `[]`

Ferdi265 replied:
done, thanks

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+            lambda x: x["mediapackage"]["media"]["track"]) or []
+
+        formats = []
+        format_types = defaultdict(lambda: defaultdict(int))
```
totally unnecessary. You can just make a set of protocols in guess_formats

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+    IE_DESC = 'tube.tugraz.at'
+    IE_NAME = 'TubeTuGraz'
+
+    _VALID_URL = r"""(?x)^
+        (?:https?://)?tube\.tugraz\.at/paella/ui/(?:
```
```suggestion
    _VALID_URL = r'''(?x)
        https?://tube\.tugraz\.at/paella/ui/(?:
```

Ferdi265 replied:
done

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+            (?P<episode>watch.html\?id=)
+        )(?P<id>[0-9a-fA-F]{8}-(?:[0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12})
+    $"""
+    _TEST = {
```
Do `_TESTS = [{}]`

Ferdi265 replied:
done

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+    _LOGIN_REQUIRED = False
+    _NETRC_MACHINE = None
+
+    _LOGIN_URL = "https://tube.tugraz.at/Shibboleth.sso/Login?target=/paella/ui/index.html"
+    _LOGIN_SUCCESS_URL = "https://tube.tugraz.at/paella/ui/index.html"
+
+    _API_SERIES = "https://tube.tugraz.at/series/series.json"
+    _API_EPISODE = "https://tube.tugraz.at/search/episode.json"
```
inline the URLs that are only used once

Ferdi265 replied:
done

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+        self._sort_formats(formats)
+
+        return {
+            "_type": "video",
```
```suggestion
```

Ferdi265 replied:
done

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+        framerate = try_get(format_info, lambda x: x["video"]["framerate"])
+        resolution = try_get(format_info, lambda x: x["video"]["resolution"])
+
+        type = str(type).replace("/delivery", "")
+        transport = str(transport).lower()
```
these are already str, no?

Ferdi265 replied:
correct, removed

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,349 @@
+                "resolution": resolution,
+            }]
+        elif transport == "hls":
+            formats = self._extract_m3u8_formats(
```
```suggestion
            formats, subs = self._extract_m3u8_formats_and_subtitles(
```
same for mpd

Ferdi265 replied:
I don't know how to return the subtitles in this case. Also, so far, none of the lectures on tube.tugraz.at I have seen have subtitles. It seems the player doesn't even support subtitles.

Ferdi265 commented:
Thank you for the detailed review, will fix it once I find time (most likely this weekend)

Ferdi265 commented:
Hi! I'm sorry that I didn't get to addressing all of your review comments, and thanks for rewriting it.

Adding the presenter/presentation info into the "MORE INFO" column is a very good idea and removes the need for my custom format ids, thanks for that.

As for the fallback formats (and "# TODO: Add test for these"):
Fallback handles episodes (not series) where metadata is only accessible for logged in users, but video data itself is still downloadable without login.

your rewrite broke that code path a bit, but that's easily fixable:
- add default (a dict with only id) for getting the episode_info dict in `TubeTuGrazIE._real_extract`
- make creator extraction robust against nonexistent creator
I fixed those two and pushed them into the branch.

I also have a test case for the fallback code path:

URL: https://tube.tugraz.at/paella/ui/watch.html?id=2df6d787-e56a-428d-8ef4-d57f07eef238
Title: "#<span></span>24 - Exakte DGL und integrierende Faktoren; lineare DGL beliebiger Ordnung"

<details>
<summary>
Info-Dict when logged in:
</summary>

```
{
  "id": "2df6d787-e56a-428d-8ef4-d57f07eef238",
  "title": "#24 - Exakte DGL und integrierende Faktoren; lineare DGL beliebiger Ordnung",
  "creator": "Philipp Sprüssel",
  "duration": 5487634,
  "series": "[503053] Mathematik B (ET)",
  "series_id": "4cb17986-bff8-4fdc-847d-3d1e4ec4981c",
  "episode": "#24 - Exakte DGL und integrierende Faktoren; lineare DGL beliebiger Ordnung",
  "webpage_url": "https://tube.tugraz.at/paella/ui/watch.html?id=2df6d787-e56a-428d-8ef4-d57f07eef238",
  "webpage_url_basename": "watch.html",
  "webpage_url_domain": "tube.tugraz.at",
  "extractor": "TubeTuGraz",
  "extractor_key": "TubeTuGraz",
  "display_id": "2df6d787-e56a-428d-8ef4-d57f07eef238",
  "fulltitle": "#24 - Exakte DGL und integrierende Faktoren; lineare DGL beliebiger Ordnung",
  "duration_string": "1524:20:34",
  "format": "p0va0br197213 - 1280x800 (presentation/delivery)+p0aa0br193850 - audio only (presenter/delivery)",
  "format_id": "p0va0br197213+p0aa0br193850",
  "ext": "mp4",
  "protocol": "http_dash_segments+http_dash_segments",
  "language": "eng",
  "format_note": "presentation/delivery+presenter/delivery",
  "filesize_approx": 274689358712,
  "tbr": 391.063,
  "width": 1280,
  "height": 800,
  "resolution": "1280x800",
  "fps": 30,
  "dynamic_range": "SDR",
  "vcodec": "avc1.4d4020",
  "vbr": 197.213,
  "acodec": "mp4a.40.2",
  "abr": 193.85,
  "asr": 44100,
  "epoch": 1657813466,
  "_type": "video"
}
```
</details>

<details>
<summary>
Info-Dict when not logged in (via fallback code path):
</summary>

```
{
  "id": "2df6d787-e56a-428d-8ef4-d57f07eef238",
  "title": "TubeTuGraz video #2df6d787-e56a-428d-8ef4-d57f07eef238",
  "webpage_url": "https://tube.tugraz.at/paella/ui/watch.html?id=2df6d787-e56a-428d-8ef4-d57f07eef238",
  "webpage_url_basename": "watch.html",
  "webpage_url_domain": "tube.tugraz.at",
  "extractor": "TubeTuGraz",
  "extractor_key": "TubeTuGraz",
  "display_id": "2df6d787-e56a-428d-8ef4-d57f07eef238",
  "format": "p0va0br197213 - 1280x800 (presentation)+p0aa0br193850 - audio only (presenter)",
  "format_id": "p0va0br197213+p0aa0br193850",
  "ext": "mp4",
  "protocol": "http_dash_segments+http_dash_segments",
  "language": "eng",
  "format_note": "presentation+presenter",
  "tbr": 391.063,
  "width": 1280,
  "height": 800,
  "resolution": "1280x800",
  "fps": 30,
  "dynamic_range": "SDR",
  "vcodec": "avc1.4d4020",
  "vbr": 197.213,
  "acodec": "mp4a.40.2",
  "abr": 193.85,
  "asr": 44100,
  "epoch": 1657813480,
  "_type": "video"
}
```
</details>

The webpage for this video should be inaccessible for non-logged-in users, but accessible for every TUGraz student. Everyone should be able to verify the "without login" testcase, and TUGraz students should be able to verify the "with login" testcase.

Since not everyone is able to verify both test cases, I have not added them to the list of test cases yet, but I think at least the fallback testcase could be added.

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,165 @@
+        episode_data = self._download_json(
+            self._API_EPISODE, video_id, query={'id': video_id, 'limit': 1}, note='Downloading episode metadata')
+
+        episode_info = traverse_obj(episode_data, ('search-results', 'result'), default={"id": video_id})
```
```suggestion
        episode_info = traverse_obj(episode_data, ('search-results', 'result'), default={'id': video_id})
```

Ferdi265 replied:
ah yes, forgot about the single quotes for that one, will fix that

pukkandan reviewed (changes requested):
> Since not everyone is able to verify both test cases, I have not added them to the list of test cases yet, but I think at least the fallback testcase could be added.

It is not expected for test cases to cover the logged in case. But pls add a test for the fallback case and also one for series extractor

Ferdi265 commented:
> > Since not everyone is able to verify both test cases, I have not added them to the list of test cases yet, but I think at least the fallback testcase could be added.
> 
> It is not expected for test cases to cover the logged in case. But pls add a test for the fallback case and also one for series extractor

Thanks. I'm adding the test case for that right now (currently trying to figure out how to run only that testcase so I can more easily get it working)

pukkandan commented:
test/test_download.py TestDownload.test_TubeTuGraz_1

Ferdi265 commented:
testcase is pushed

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -117,20 +117,35 @@ class TubeTuGrazIE(TubeTuGrazBaseIE):
+                'id': '2df6d787-e56a-428d-8ef4-d57f07eef238',
+                'ext': 'mp4',
+                'title': '2df6d787-e56a-428d-8ef4-d57f07eef238',
+                'episode': None,
+                'series': None,
+                'creator': None,
+                'duration': None,
+                'series_id': None,
```
```suggestion
```

Ferdi265 replied:
removed those fields

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -45,7 +45,7 @@ def _extract_episode(self, episode_info):
         return {
             'id': id,
-            'title': title,
+            'title': title or id,
```
If you cannot extract title, just leave it as None
```suggestion
            'title': title,
```

Ferdi265 replied:
if I leave it as None, the testcase fails because the "no title given, generic title used" warning seems to be an error in test mode

pukkandan replied:
add it to `expected_warnings` in the test

Ferdi265 replied:
ahh! did not know that field existed, thanks

Ferdi265 replied:
done

pukkandan commented on yt_dlp/extractor/tubetugraz.py:
```diff
@@ -0,0 +1,180 @@
+                'creator': None,
+                'duration': None,
+                'series_id': None,
+            }
+        }
```
```suggestion
            },
            'expected_warnings': ['no title given']
        },
```
like this

pukkandan commented:
still waiting for a series test

Ferdi265 commented:
> still waiting for a series test

I'll check if there are any series accessible to non-logged in users. If I find one, I'll add that as a test

Ferdi265 commented:
series test added.
